subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/forsubgraph/SubgraphVersionConstraintsFeatureInteractionIntegrationTest.groovy

```diff
@@ -0,0 +1,325 @@
+                                }
+                            }
+                        }
+                    }
+                }
+                conf('org:bar:1.0')
+            }           
+        """
+
+        when:
+        repositoryInteractions {
+            'org:foo:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+            'org:bar:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+            'org:baz:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+        }
+        run ':checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(':', ':test:') {
+                module('org:bar:1.0') {
+                    module('org:baz:1.0').byRequest()
+                    module('org:foo:1.0') {
+                        edge('org:baz:2.0', 'org:baz:1.0').byAncestor()
+                    }
+                }
+            }
+        }
+    }
+
+    def "an ancestor provided versions is not a version conflict"() {
+        given:
+        repository {
+            'org:foo:1.0'()
+            'org:foo:2.0'()
+            'org:bar:1.0' {
+                dependsOn 'org:foo:2.0'
+            }
+        }
+
+        buildFile << """
+            configurations.conf.resolutionStrategy.failOnVersionConflict()
+            dependencies {
+                constraints {
+                    conf('org:foo:1.0') {
+                       version { forSubgraph() }
+                    }
+                }
+                conf('org:bar:1.0')
+            }           
+        """
+
+        when:
+        repositoryInteractions {
+            'org:foo:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+            'org:bar:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+        }
+        run ':checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(':', ':test:') {
+                constraint('org:foo:1.0').byConstraint()
+                module('org:bar:1.0') {
+                    edge('org:foo:2.0', 'org:foo:1.0').byAncestor()
+                }
+            }
+        }
+    }
+
+    def "can force a version over an ancestor provided version"() {
+        given:
+        repository {
+            'org:bar:1.0'()
+            'org:bar:2.0'()
+        }
+
+        settingsFile << "\ninclude 'foo'"
+        buildFile << """
+            project(':foo') {
+                configurations.create('conf')
+                artifacts { add('conf', file('foo.jar')) }
+                dependencies { 
+                    conf('org:bar:2.0') { force = true }
+                }
+            }
+            dependencies {
+                constraints {
+                    conf('org:bar:1.0') {
+                       version { forSubgraph() }
```

Comment:
I think this is the right semantics: _force_ forces the version, no matter what else. Which does not mean one _should_ do this. This test only documents how it is.

I think we should make clear in the docs that _force_ is only something for a local project (if you are building an application). Then you may use it, because it's "simple". For libraries, you should not use it at all. (And we also do not publish it therefore.)

If the `forSubgraph` is good enough for _all_ use cases, we can consider deprecating _force_ completely at some point...